cpp/src/io/utilities/hostdevice_span.hpp

```diff
@@ -33,31 +34,18 @@ class hostdevice_span {
   hostdevice_span(hostdevice_span const&) = default;  ///< Copy constructor
   hostdevice_span(hostdevice_span&&)      = default;  ///< Move constructor
 
-  hostdevice_span(T* cpu_data, T* gpu_data, size_t size)
-    : _size(size), _device_data(gpu_data), _host_data(cpu_data)
+  hostdevice_span(host_span<T> host_data, T* device_data)
+    : _host_data{host_data}, _device_data{device_data}
   {
   }
 
-  /// Constructor from container
-  /// @param in The container to construct the span from
-  template <typename C,
-            // Only supported containers of types convertible to T
-            std::enable_if_t<std::is_convertible_v<
-              std::remove_pointer_t<decltype(std::declval<C&>().host_ptr())> (*)[],  // NOLINT
-              T (*)[]>>* = nullptr>                                                  // NOLINT
-  constexpr hostdevice_span(C& in) : hostdevice_span(in.host_ptr(), in.device_ptr(), in.size())
```

Comment:
This was only used to construct const versions of existing `hostdevice_span`s, So I made it more concrete. We still can create a `hostdevice_span` from `hostdevice_vector`, and that's the only container we convert to this kind of span.